test/loadtime/report/report.go

```diff
@@ -0,0 +1,121 @@
+package report
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/tendermint/tendermint/test/loadtime/payload"
+	"github.com/tendermint/tendermint/types"
+	"gonum.org/v1/gonum/stat"
+)
+
+// BlockStore defines the set of methods needed by the report generator from
+// Tendermint's store.Blockstore type. Using an interface allows for tests to
+// more easily simulate the required behavior without having to use the more
+// complex real API.
+type BlockStore interface {
+	Height() int64
+	Base() int64
+	LoadBlock(int64) *types.Block
+}
+
+// Report contains the data calculated from reading the timestamped transactions
+// of each block found in the blockstore.
+type Report struct {
+	Max, Min, Avg, StdDev time.Duration
+
+	// ErrorCount is the number of parsing errors encountered while reading the
+	// transaction data. Parsing errors may occur if a transaction not generated
+	// by the payload package is submitted to the chain.
+	ErrorCount int
+
+	// All contains all data points gathered from all valid transactions.
+	All []time.Duration
+}
+
+// GenerateFromBlockStore creates a Report using the data in the provided
+// BlockStore.
+func GenerateFromBlockStore(s BlockStore) (Report, error) {
+	type payloadData struct {
+		l   time.Duration
+		err error
+	}
+	type txData struct {
+		tx []byte
+		bt time.Time
+	}
+
+	const poolSize = 16
+
+	txc := make(chan txData)
+	pdc := make(chan payloadData, poolSize)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(poolSize)
+	for i := 0; i < poolSize; i++ {
+		go func() {
+			defer wg.Done()
+			for b := range txc {
+				p, err := payload.FromBytes(b.tx)
+				if err != nil {
+					pdc <- payloadData{err: err}
+					continue
+				}
+				l := b.bt.Sub(p.Time.AsTime())
```

Comment:
We agreed that, for this scheme to work well, we'll need to configure NTP on nodes involved.
In order to (try to) catch the fact that clocks are not synchronized, could you add some (rough) sanity checks here? For instance:
* l should be positive
* I should be less than, say, 120 seconds

These could also be done by inspecting the results provided by the report (say, the script that calls `report`). Not sure what's best here.

Reply:
> the resulting duration should be positive

I think it may make sense to just report a count of negative durations. I doubt we want to error in those cases since the data is still perhaps interesting. Does this seem reasonable?

Reply:
Thinking a bit more about this, probably a single flag in the report (set to true if _any_ delta is negative or ridiculously big) should suffice.
If we have evidence the the clocks aren't synchronized (because we forgot, or NTP daemon crashed, or was accidentally stopped in one of the machines), the simplest course of action to me would be to repeat the test from scratch with a good NTP setup, rather than trying to extract info from a report with skewed measurements.

In short, a count of negative (or very big) durations would be good, but we can also have just a general flag ("clocks likely not in sync").

Reply:
> ridiculously big

We have the `Min` and `Max` so users can determine this on their own. 

> but we can also have just a general flag ("clocks likely not in sync").

I'd prefer to not try to be much smarter than we need to be at the possible expense of clarity. Adding such a flag is doing the work of interpreting the data for the user which I don't think this tool should be doing. IMO, it should report the data in a manner that helps the user draw conclusions about what happened during the experimental run. It _should_ be clear enough from the outputs given (large number of negative durations, or huge values for Max and Min) that the clocks are not in sync.